triage/summarize/cluster.go

```diff
@@ -60,17 +67,81 @@ func clusterLocal(failuresByTest failuresGroup) nestedFailuresGroups {
 		return clustered
 	}
 
-	numFailures := 0
+	numFailures := 0 // The number of failures processed so far
 	start := time.Now()
 	logInfo("Clustering failures for %d unique tests...", len(failuresByTest))
 
-	// Look at tests with the most failures first.
-	for n, pair := range failuresByTest.sortByNumberOfFailures() {
-		numFailures += len(pair.Failures)
-		logInfo("%4d/%4d tests, %5d failures, %s", n+1, len(failuresByTest), len(pair.Failures), pair.Key)
-		clustered[pair.Key] = clusterTest(pair.Failures)
+	/*
+		Since local clustering is done within each test, there is no interdependency among the clusters,
+		and the clustering process can be easily multithreaded.
+
+		One goroutine will push to a work queue, and one goroutine will read from a done queue.
+		The specified number of (or, if none are specified, NumCPU-1) goroutines will act as workers.
+		Each worker will pull from the work queue and perform a clustering operation, and then push
+		the result to the done queue.
+	*/
+
+	var workerWG sync.WaitGroup
+	var doneWG sync.WaitGroup // Prevents this function from returning results before they are all written to clustered
+
+	numWorkers := threads
+	if threads <= 0 {
+		numWorkers = runtime.NumCPU() - 1 // Dedicate one CPU to pushing to the work queue/pulling from the done queue/main thread
```

Comment:
suggestion: I usually recommend hauling references to globals up the stack, it makes it easier to unit test etc. later

I would put this as the default for a flag that allows tuning this.

also: what if NumCPU is 1?

Reply:
I set the flag default as max of NumCPU-1 or 1, and then also set a (I think sensible) default of 1 for the function if numWorkers is passed in as <= 0.